### PR TITLE
create all statefulsets immediately

### PR DESCRIPTION
**Pull request by gregwebs**
# TODO

- [ ] run stability tests
- [ ] decide on renaming wait-for-pd

As per #469 this creates statefulsets and pending pods immediately to help autoscalers scale more quickly.

Design doc is [here](https://docs.google.com/document/d/1LFlGEy2NRwtmbwUb_hS9_B7POIseBFUsOSyhezf1RXg)

A blocking init container wait-for-pd is used to prevent TiKV and TiDB from immediately starting up. This code calls the PD API directly. For TiDB it will call the new status is_initialized API if it is available.

Error handling is also generally improved in the member managers: as a side effect this may fix bugs where the controller sync loop is not re-ran during errors.

The PD API was updated to 3.0 which seems to be backwards compatible for our usage (more testing needed).

### Testing notes

I have created 50+ clusters with the operator that has the changes from the first commit. It has the desired effect for autoscaling. The second commit is still being tested.

### What is changed and how does it work?

* When PD is not available for TiKV, or TiKV is not available for TiDB, go ahead and create the statefulset.
* Don't break from the update function immediately when a RequeueError is seen.

### Check List

Tests

 - Unit test: using existing tests
 - E2E test: using existing e2e tests
 - Stability test: ?
 - Manual test: create single clusters of various sizes (multiple cluster creation often runs into issues with the current operator). Test with pd/tikv/tidb version 3.0.0-rc1 and version 3.0.0

### Side effects

Increased image size
Originally this change meant shipping another 50M binary. However, overall image size is smaller since this PR was originally started. This work was merged in other Pull requests:
* #618 Code was re-arragned so that wait-for-pd minimized its imports to just the interaction with the PD API and logging without importing all of Kubernetes. This reduced the binary size by 50%. We could do the same for the discovery service.
* #592 Removal of debugger symbols reduced the size of all the binaries

Pending Pod warnings
* May surprise existing users that pre-provision capacity and monitor for pending pods. Alerts that alert for pending pods could get triggered.

Upgrade procedure
* When upgrading the tidb-operator image from before this change one must first upgrade the tidb-operator helm chart templates. This is expected, but it could surprise someone since it often isn't necessary.

The PD API was updated to 3.0 which seems to be backwards compatible for our usage. We will probably want to see some additional testing here.

### Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?

Yes, to the extent users look at pod creation (which we have in our tutorials).

 ```release-note
Create TiKV/TiDB pending pods immediately to help autoscalers scale more quickly. They wait for their dependencies (e.g. PD) to be ready with an Init container, so they will show up as in state Init while waiting.
 ```

### CODE REVIEW

**Comment by gregwebs** [comment]
This is working well now. I just need to go through the tests. There are some additional things on this branch that I am getting merged in other PRs right now and then I will force push a rebased version of this.

* gcp terraform
* error message improvements
* binary build size reduction

**Comment by gregwebs** [comment]
I saw an issue where TiKV continually fails:

```
[2019/06/20 23:49:42.564 +00:00] [FATAL] [tikv-server.rs:273] ["failed to start node: StringError(\"[src/server/node.rs:171]: cluster ID mismatch, local 6704759663323983770 != remote 6704764185849198648, you are trying to connect to another cluster, please reconnect to the correct PD\")"]
```

This probably means I need to change the wait-for-pd logic to be more like `PDIsAvailable`

**Comment by tennix** [comment]
You are reusing old data for the new cluster.

**Comment by gregwebs** [comment]
@tennix I was destroying and creating clusters so it would be expected for a node to be re-used by a pod of a new cluster. Is this expected behavior then? Note that I deleted the PVC. Am I supposed to manually wipe the PV?

**Comment by tennix** [comment]
Did you use local ssd disk or persistent disk?
I suppose you are using the local ssd disk. So if you delete the PVC, I don't think the PV will be reused. If the PV reclaim policy is `Delete`, then the old PV and data will be deleted by the provisioner, after that a new PV will be provisioned. If the policy is `Retain`, then the PV will be in `Released` state, it will not be used.

**Comment by gregwebs** [comment]
ah, it seems the problem is that I wasn't setting the reclaim policy to Delete.
Because this PR creates pods immediately, node rescheduling is now more possible in my setup where before an empty node would have been automatically deleted.

**Comment by gregwebs** [comment]
I will fix a couple failing tests now. I am ready to start review and get suggestions on test cases to write.

File: pkg/apis/pdapi/pdapi.go [hunk 1]
```diff
@@ -0,0 +1,797 @@
+// Copyright 2018 PingCAP, Inc.
```
Comment (cofyc) [inline pkg/apis/pdapi/pdapi.go]: `pkg/apis` is a directory to store Kubernetes API types. IMO it's not a good idea to put PD controller client here.

File: pkg/controller/pd_control.go [hunk 2]
```diff
@@ -14,802 +14,18 @@
 package controller
```
Comment (cofyc) [inline pkg/controller/pd_control.go]: how about focusing on design change in this PR and optimize later?

File: pkg/apis/pdapi/pdapi.go (hunk 1 above)
Comment (gregwebs) [reply pkg/apis/pdapi/pdapi.go]: Where would a good place be for this?
I don't think of this as a PD controller but instead just the PD API interaction. Logic to "control" PD is elsewhere. The code take types from this project, serializes them for the API, sends a request, and then convert the response into a type for this project. The file is renamed now to pdapi. There is an interface PDControlInterface that I would also rename, but I tried to minimize my changes to only what is required.

File: pkg/controller/pd_control.go (hunk 2 above)
Comment (gregwebs) [reply pkg/controller/pd_control.go]: okay, here is the PD API change: https://github.com/pingcap/tidb-operator/pull/618

**Comment by gregwebs** [comment]
The usage of PD APIs in wait-for-pd should be updated. For TiDB 3.0 the TiDB component should check the PD api cluster_status for the is_initialized flag as per #586

**Comment by gregwebs** [comment]
One piece of feedback I got it is that the name "wait-for-pd" may not accommodate future changes well. "tidb-init" is a very generic name, but I think too generic. An example given is password initialization: we might move that to an init container, but I think it will be best off as a separate executable.
My suggestion is "wait-for-deps" would help accommodate the fact that we are actually waiting for TiKV replication. That would also work with the example of putting the waiting for the discovery service that PD does into the same binary. "deps" in "wait-for-deps" is not the most clear name. "services" might be better.

**Comment by gregwebs** [comment]
Closes #586.

File: cmd/wait-for-pd/main.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+		waitFunction = waitForLeaderFunc
+	} else {
+		glog.Fatalf("Expected either the flag --initialization or --leader")
+	}
+
+	for {
+		if waitFunction() {
+			break
+		}
+		time.Sleep(oneSecond)
```
Comment (cofyc) [inline cmd/wait-for-pd/main.go]: nit: `time.Second`

File: cmd/wait-for-pd/main.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	if printVersion {
+		version.PrintVersionInfo()
+		os.Exit(0)
+	}
+	version.LogVersionInfo()
+
+	logs.InitLogs()
+	defer logs.FlushLogs()
+
+	oneSecond, err := time.ParseDuration("1s")
```
Comment (cofyc) [inline cmd/wait-for-pd/main.go]: configurable in the future?

Comment (cofyc) [reply cmd/wait-for-pd/main.go]: I think we can keep it simple at first.

**Comment by cofyc** [comment]
target 1.1 release, what do you think?

**Comment by gregwebs** [comment]
@cofyc that decision doesn't seem to be up to me. However, this PR does fix some problematic error handling which can be seen in the test cases.

**Comment by gregwebs** [comment]
I see an issue where the operator will expand the tidb statefulset when it is waiting in init for a long time

```
demo-discovery-5bb58f7bb5-wklmq   1/1     Running    0          18m
demo-monitor-6f9c9fdffb-59bkp     1/1     Running    0          18m
demo-pd-0                         1/1     Running    0          18m
demo-pd-1                         1/1     Running    0          18m
demo-pd-2                         1/1     Running    1          18m
demo-tidb-0                       0/2     Init:0/1   0          18m
demo-tidb-1                       0/2     Init:0/1   0          18m
demo-tidb-2                       0/2     Init:0/1   0          9m23s
demo-tidb-3                       0/2     Init:0/1   0          9m23s
demo-tidb-4                       0/2     Init:0/1   0          4m23s
demo-tikv-0                       0/1     Pending    0          18m
demo-tikv-1                       1/1     Running    0          18m
demo-tikv-2                       1/1     Running    0          18m
```

**Comment by gregwebs** [comment]
When this came out of the init state the extra tidb pods did get terminated.

File: cmd/wait-for-pd/main.go (hunk 4 above)
Comment (gregwebs) [reply cmd/wait-for-pd/main.go]: yes, configurable in the future

File: cmd/wait-for-pd/main.go (hunk 3 above)
Comment (gregwebs) [reply cmd/wait-for-pd/main.go]: We may want the time to be configurable in the future

**Comment by gregwebs** [comment]
okay, I added some code to handle that case

File: pkg/manager/member/tidb_member_manager.go [hunk 5]
```diff
@@ -151,8 +154,12 @@ func (tmm *tidbMemberManager) syncTiDBStatefulSetForTidbCluster(tc *v1alpha1.Tid
 		tc.Status.TiDB.StatefulSet = &apps.StatefulSetStatus{}
 		return nil
 	}
-
-	if !tc.TiKVIsAvailable() {
+	initialized, err := controller.GetPDClient(tmm.pdControl, tc).GetClusterInitialized()
```
Comment (gregwebs) [inline pkg/manager/member/tidb_member_manager.go]: Doing an API call here may slow down the sync loop. Previously it checked the status. Perhaps this API call should be done elsewhere and set a status field that is checked here?

File: pkg/manager/member/member.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+		},
+		{
+			Name:  "CLUSTER_NAME",
+			Value: tcName,
+		},
+	}
+
+	return corev1.Container{
+		Name:    "wait-for-pd",
+		Image:   operatorImage,
```
[the remark this answers is not in the document]
Comment (gregwebs) [reply pkg/manager/member/member.go]: Good point: this function should only be used on create, but not update.

File: cmd/wait-for-pd/main.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+package main
+
+import (
+	"github.com/golang/glog"
+	"github.com/pingcap/tidb-operator/pkg/pdapi"
+	"github.com/pingcap/tidb-operator/version"
+	"k8s.io/apiserver/pkg/util/logs"
+
+	"flag"
```
[the remark this answers is not in the document]
Comment (gregwebs) [reply cmd/wait-for-pd/main.go]: gofmt seems fine with that. I moved it above now though.

File: pkg/manager/member/member.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gregwebs) [reply pkg/manager/member/member.go]: I added the change only on create functionality

File: pkg/manager/member/tidb_member_manager.go (hunk 5 above)
Comment (gregwebs) [reply pkg/manager/member/tidb_member_manager.go]: I moved this into the PD status: it will be easier to create a PD API that reduces our API calls there.

**Comment by aylei** [comment]
@gregwebs Cut of `release-1.1` branch is targeted at December 31th, could you please confirm whether this feature is expected to be releases in 1.1 or should be re-targeted to 1.2? Thanks!

**Comment by gregwebs** [comment]
Probably needs to be 1.2

**Comment by aylei** [comment]
Okay, removed from v1.1 milestone for now, feel free to add it back if this is necessary to land in v1.1

**Pull request closed without merging**